File: M3 Intermediate DSA/src/L10_Interview_Problems_on_Arrays/Assignment4.java
package L10_Interview_Problems_on_Arrays;

public class Assignment4 {

  public int majorityElement(int[] A) {
    int N = A.length;
    int ans = A[0];
    int freq = 1;

    for (int index= 1; index < N; index++) {
      if (freq == 0) {
        ans = A[index];
        freq = 1;
      } else if (A[index] == ans) {
        freq++;
      } else  {
        freq--;
      }
    }

    return ans;
  }

}

File: M3_Intermediate_DSA/L8_Interview_Problems_on_Arrays/assignment4.go
package main

func MajorityElement(A []int) int {
	freq := 1
	currEle := A[0]

	for i := 1; i < len(A); i++ {
		if freq == 0 {
			currEle = A[i]
			freq = 1
		} else if currEle == A[i] {
			freq++
		} else {
			freq--
		}
	}

	return currEle
}
